EIPS/eip-6774.md

```diff
@@ -0,0 +1,262 @@
+- Company Benefits: A company could be giving out benefits to its employees in a closed environment its controls (hence being able to freely add in or remove redeemable benefits).
+
+NSTs propose a way to enforce users to perform barter of tokens between two authorized tokens in order to maintain a value equivalence in the exchange and so reduce the risk of speculation associated with an one way transfer. NSTs would provide a solution for transferring non-valuable tokens between projects based on their true value instead of their perceived/ speculated value without leveraging a third party.
+
+## Specification
+
+The key words "MUST", "REQUIRED", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Overview
+
+An approval mechanism is REQUIRED before performing barters between two tokens, belonging to the same or different NST contracts (see [Permissonless barter](#permissionless-barter)).
+
+Every contract compliant with this EIP MUST implements this transfer mechanism:
+
+1. UserA signs a message taking up the terms of the barter
+2. UserB accepts and concludes the barter by calling the `barter` function
+
+```mermaid
+sequenceDiagram
+    participant UserA
+    note right of UserA: BarterTerms:<br/>give: <br/>NST A<br/>TokenID:2<br/>ask:<br/>NST B<br/>TokenID:5<br/>
+    participant UserB
+    UserA-->>UserA: Sign barter terms
+    note right of UserA: Message:<br/>- BarterTerms<br/>- UserA signature
+    UserA-->>UserB: Send off-chain barter message
+    UserB->>+NST B: call barter(message)
+    NST B->>+NST A: call transferFor(message)
+    NST A->>NST A: check UserA signature and increase nonce
+    NST A->>-UserB: transfer NST A tokenID: 2
+    NST B->>-UserA: transfer NST B tokenID: 5
+```
+
+### NST
+
+#### Methods
+
+In the following, the `BarterTerms` struct are defined as follows:
+
+```solidity
+struct Component {
+    address tokenAddr;
+    uint256 tokenId;
+}
+
+struct BarterTerms {
+    Component bid;
+    Component ask;
+    uint256 nonce;
+    address owner;
+    uint48 deadline;
+}
+```
+
+See [Define the barter terms](#define-the-barter-terms) section for precisions on this struct.
+
+##### transferFor
+
+Verifies signature following the [EIP-712](./eip-712.md) procedure, increases the nonce of `data.owner` and transfers the `data.bid.tokenId` to `to` account.
+
+This function MUST be called only by authorized NST contracts, the call MUST occur when users are calling the `barter` function.
+
+NOTE: using `ECDSA.sol`(`v4.7.3` at least) and `EIP-712.sol` libraries from OpenZeppelin are very RECOMMENDED for signature verifications, verification only based on `ecrecover` SHOULD NOT be used (see the Solidity documentation, Mathematical and Cryptographic Functions).
```

Comment:
```suggestion
```

You can make a note of ECDSA problems in the security considerations section, but generally we prefer to avoid implementation details in the specification section.